src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,249 @@
-        A.evar s)
-      l
-    |> A.pexp_tuple
-
-  let rec nest ~loc ~pair ~triple ~quad = function
-    | Quad (a, b, c, d) -> [%expr [%e quad] [%e a] [%e b] [%e c] [%e d]]
-    | Triple (a, b, c) -> [%expr [%e triple] [%e a] [%e b] [%e c]]
-    | Pair (a, b) ->
-        [%expr
-          [%e pair]
-            [%e nest ~loc ~pair ~triple ~quad a]
-            [%e nest ~loc ~pair ~triple ~quad b]]
-    | Elem a -> a
-
-  let to_gen ~loc t =
-    let pair = [%expr QCheck.Gen.pair] in
-    let triple = [%expr QCheck.Gen.triple] in
-    let quad = [%expr QCheck.Gen.quad] in
-    nest ~loc ~pair ~triple ~quad t
-
-  let to_obs ~loc t =
-    let pair = [%expr QCheck.Observable.pair] in
-    let triple = [%expr QCheck.Observable.triple] in
-    let quad = [%expr QCheck.Observable.quad] in
-    nest ~loc ~pair ~triple ~quad t
-
-  let to_pat ~loc t =
-    let fresh_id =
-      let id = ref 0 in
-      fun () ->
-        let x = !id in
-        let () = id := x + 1 in
-        Printf.sprintf "gen%d" x
-    in
-    let (module A) = Ast_builder.make loc in
-    let rec aux = function
-      | Quad (_, _, _, _) ->
-          let a = A.pvar @@ fresh_id () in
-          let b = A.pvar @@ fresh_id () in
-          let c = A.pvar @@ fresh_id () in
-          let d = A.pvar @@ fresh_id () in
-          [%pat? [%p a], [%p b], [%p c], [%p d]]
-      | Triple (_, _, _) ->
-          let a = A.pvar @@ fresh_id () in
-          let b = A.pvar @@ fresh_id () in
-          let c = A.pvar @@ fresh_id () in
-          [%pat? [%p a], [%p b], [%p c]]
-      | Pair (a, b) ->
-          let a = aux a in
-          let b = aux b in
-          [%pat? [%p a], [%p b]]
-      | Elem _ -> A.pvar @@ fresh_id ()
-    in
-    aux t
-end
+    [f] can be used to transform tuples, for instance:
+    {[
+    type t = Foo of int * int
+    ]}
 
-let map ~loc pat expr gen =
-  [%expr QCheck.Gen.map (fun [%p pat] -> [%e expr]) [%e gen]]
+    Without [f]:
+    {[
+    let gen = QCheck.Gen.(map (fun (x, y) -> (x, y)) (pair int int))
+    ]}
 
+    With [f], building Foo:
+    {[
+    let gen = QCheck.Gen.(map (fun (x, y) -> Foo (x, y)) (pair int int))
+    ]}
+*)
 let tuple ~loc ?(f = fun x -> x) tys =
   let tuple = Tuple.from_list tys in
   let gen = Tuple.to_gen ~loc tuple in
   let expr = Tuple.to_expr ~loc tuple |> f in
   let pat = Tuple.to_pat ~loc tuple in
-  map ~loc pat expr gen
+  G.map ~loc pat expr gen
+
+(** [record loc gens ?f label_decls] transforms [gens] and [label_decls] to
+    a record generator.
+
+    Similarly to {!tuple}, we can use [f] to transform records, for instance:
+    {[
+    type t = Foo of { left : int; right : int }
+    ]}
+
+    Without [f]:
+    {[
+    let gen = QCheck.Gen.(map (fun (x, y) -> {left = x; right = y}) (pair int int))
+    ]}
 
+    With [f], building Foo:
+    {[
+    let gen = QCheck.Gen.(map (fun (x, y) -> Foo {left = x; right = y}) (pair int int))
+    ]}
+
+*)
 let record ~loc ~gens ?(f = fun x -> x) xs =
```

Comment:
rename to `gen_record`?

Reply:
Done in d527aa353bf3f971791ff1a3e112d8d2d24179d1